pkg/tasks/probes.go

```diff
@@ -103,7 +104,32 @@ func runProbes(s *state.State) error {
 	}
 
 	if s.LiveCluster.IsProvisioned() {
-		return investigateCluster(s)
+		if err := investigateCluster(s); err != nil {
+			return err
+		}
+	}
+
+	for _, host := range s.Cluster.ControlPlane.Hosts {
+		switch host.OperatingSystem {
+		case kubeoneapi.OperatingSystemNameFlatcar, kubeoneapi.OperatingSystemNameCoreOS:
+			s.Cluster.ContainerRuntime.Docker = &kubeoneapi.ContainerRuntimeDocker{}
+		}
+	}
+
+	switch {
+	case s.Cluster.ContainerRuntime.Containerd != nil:
+		return nil
+	case s.Cluster.ContainerRuntime.Docker != nil:
+		return nil
+	}
+
+	gteKube121Condition, _ := semver.NewConstraint(">= 1.21")
+
+	switch {
+	case gteKube121Condition.Check(s.LiveCluster.ExpectedVersion) && !s.LiveCluster.IsProvisioned():
+		s.Cluster.ContainerRuntime.Containerd = &kubeoneapi.ContainerRuntimeContainerd{}
+	default:
+		s.Cluster.ContainerRuntime.Docker = &kubeoneapi.ContainerRuntimeDocker{}
 	}
```

Comment:
This logic is not going to work for newly created 1.21 clusters.

For example:
* User creates 1.21.0 cluster
* The cluster gets provisioned with containerd
* User attempts to upgrade the cluster to 1.21.1+
* This logic would default to Docker because the cluster is provisioned, so `!s.LiveCluster.IsProvisioned()` would return false
* KubeOne would install Docker and eventually break the cluster

Reply:
That's only seems like so, because you've missed the [static defaulting](https://github.com/kubermatic/kubeone/pull/1205/files#diff-8e3eafbecc2f7f4dbfd402187a37180eb206b8fe2128ddf3a60bb7fd2717bf07R135-R137)

Reply:
OK, I've misread this one. Added some live checks